### PR TITLE
interfaces/mir: allow Wayland socket and non-root sockets

### PR DESCRIPTION
**Pull request by gerboland**
Mir 0.28 and greater implements the Wayland protocol, and needs to create its socket. 

Mir running as non-root will create those sockets in $XDG_RUNTIME_DIR which is typically /run/users/*, so allow that too.

### CODE REVIEW

File: interfaces/builtin/mir.go [hunk 1]
```diff
@@ -49,6 +49,9 @@ capability sys_tty_config,
 
 /{dev,run}/shm/\#* mrw,
 /run/mir_socket rw,
+/run/user/[0-9]*/mir_socket rw,
```
Comment (jdstrand) [inline interfaces/builtin/mir.go]: I'm surprised we didn't notice this access before since we allow it in mirConnectedPlugAppArmor.

File: interfaces/builtin/mir.go [hunk 2]
```diff
@@ -49,6 +49,9 @@ capability sys_tty_config,
...
+/run/wayland* rw,
+/run/user/[0-9]*/wayland* rw,
```
Comment (jdstrand) [inline interfaces/builtin/mir.go]: I think the wayland interface should be updated to be implicit on classic and providable as a snap on non-classic (similar to how avahi-observe was recently converted) with these rules in the waylandPermanentSlotAppArmor policy. Then the mir snap would slots both mir and wayland.

File: interfaces/builtin/mir.go (hunk 1 above)
Comment (gerboland) [reply interfaces/builtin/mir.go]: Yep, I found that funny too

File: interfaces/builtin/mir.go (hunk 2 above)
Comment (gerboland) [reply interfaces/builtin/mir.go]: Yep I initially tried the Mir snap providing the wayland interface, but it failed as it was not the core snap. I agree your proposal is much clearer, I'll do that then

File: interfaces/builtin/wayland.go [hunk 3]
```diff
@@ -19,32 +19,160 @@
...
 
 const waylandBaseDeclarationSlots = `
   wayland:
     allow-installation:
       slot-snap-type:
+        - app
         - core
+    deny-auto-connection: true
+    deny-connection:
+      on-classic: false
```
Comment (jdstrand) [inline interfaces/builtin/wayland.go]: You want this for auto-connection:
```
deny-auto-connection:
  on-classic: false
```

File: interfaces/builtin/wayland.go [hunk 4]
```diff
@@ -19,32 +19,160 @@
...
+	old := "###SLOT_SECURITY_TAGS###"
+	var new string
+	if release.OnClassic {
+		// If we're running on classic Wayland will be part
+		// of the OS snap and will run unconfined.
+		new = "unconfined"
+	} else {
+		new = slotAppLabelExpr(slot)
+	}
+	snippet := strings.Replace(waylandConnectedPlugAppArmor, old, new, -1)
```
Comment (jdstrand) [inline interfaces/builtin/wayland.go]: This code will change if the unix rule goes away.

File: interfaces/builtin/wayland.go [hunk 5]
```diff
@@ -19,32 +19,160 @@
...
+	return nil
+}
+
+func (iface *waylandInterface) AppArmorConnectedSlot(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	if !release.OnClassic {
+		old := "###PLUG_SECURITY_TAGS###"
+		new := plugAppLabelExpr(plug)
+		snippet := strings.Replace(waylandConnectedSlotAppArmor, old, new, -1)
+		spec.AddSnippet(snippet)
+	}
```
Comment (jdstrand) [inline interfaces/builtin/wayland.go]: This code will change if the unix rule goes away.

File: interfaces/builtin/wayland.go [hunk 6]
```diff
@@ -19,32 +19,160 @@
...
+
+const waylandPermanentSlotAppArmor = `
+# Description: Allow operating as a Wayland display server. This gives privileged access
+# to the system.
+
+# needed since Wayland is the display server, to configure tty devices
+capability sys_tty_config,
+/dev/tty[0-9]* rw,
+
+/{dev,run}/shm/\#* mrw,
```
[the remark this answers is not in the document]
Comment (gerboland) [reply interfaces/builtin/wayland.go]: The TTY stuff is needed by weston, but not shm

File: interfaces/builtin/wayland.go [hunk 7]
```diff
@@ -19,32 +19,160 @@
...
+
+/run/wayland-[0-9]* rw,
+/run/user/[0-9]*/wayland-[0-9]* rw,
+
+# XWayland requires access to this
+/run/xwayland-shared-* rw,
```
[the remark this answers is not in the document]
Comment (gerboland) [reply interfaces/builtin/wayland.go]: Nope, my bad, removed

File: interfaces/builtin/wayland.go [hunk 8]
```diff
@@ -19,32 +19,160 @@
...
+const waylandPermanentSlotSecComp = `
+# Description: Allow operating as a Wayland server. This gives privileged access
+# to the system.
+# Needed for server launch
+bind
+listen
+# Needed by server upon client connect
+accept
+accept4
+shmctl
```
[the remark this answers is not in the document]
Comment (gerboland) [reply interfaces/builtin/wayland.go]: Just accept4 by weston

File: interfaces/builtin/wayland.go [hunk 9]
```diff
@@ -19,32 +19,160 @@
...
+# for udev
+socket AF_NETLINK - NETLINK_KOBJECT_UEVENT
+`
+
+const waylandConnectedSlotAppArmor = `
+# Description: Permit clients to use Wayland
+unix (receive, send) type=seqpacket addr=none peer=(label=###PLUG_SECURITY_TAGS###),
```
[the remark this answers is not in the document]
Comment (gerboland) [reply interfaces/builtin/wayland.go]: You're right, I have removed it

File: interfaces/builtin/wayland.go [hunk 10]
```diff
@@ -19,32 +19,160 @@
...
 `
 
 const waylandConnectedPlugAppArmor = `
-# Description: Can access compositors supporting the wayland protocol
+# Description: Permit clients to connect to compositors supporting the Wayland protocol
+unix (receive, send) type=seqpacket addr=none peer=(label=###SLOT_SECURITY_TAGS###),
```
[the remark this answers is not in the document]
Comment (gerboland) [reply interfaces/builtin/wayland.go]: Yep, removed

File: interfaces/builtin/wayland.go (hunk 4 above)
Comment (gerboland) [reply interfaces/builtin/wayland.go]: Done

File: interfaces/builtin/wayland.go (hunk 5 above)
Comment (gerboland) [reply interfaces/builtin/wayland.go]: and done

**Comment by gerboland** [comment]
Ok, iteration 2 ready. I've whipped up a quick Weston snap to test it with:
https://code.launchpad.net/~gerboland/+git/weston-snap/+ref/master

File: interfaces/builtin/wayland.go [hunk 11]
```diff
@@ -19,32 +19,127 @@
...
 const waylandSummary = `allows access to compositors supporting wayland protocol`
 
 const waylandBaseDeclarationSlots = `
   wayland:
     allow-installation:
       slot-snap-type:
+        - app
         - core
+    deny-auto-connection:
+      on-classic: false
```
Comment (jdstrand) [inline interfaces/builtin/wayland.go]: This wasn't fixed in the manner needed. Please adjust to be:
```
  wayland:
    allow-installation:
      slot-snap-type:
        - app
        - core
    deny-connection:
      on-classic: false
    deny-auto-connection:
      on-classic: false
```

File: interfaces/builtin/wayland.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (jdstrand) [reply interfaces/builtin/wayland.go]: Right, was only talking about the shm access.

File: interfaces/builtin/wayland.go [hunk 12]
```diff
@@ -19,32 +19,127 @@
...
+`
+
+const waylandPermanentSlotSecComp = `
+# Description: Allow operating as a Wayland server. This gives privileged access
+# to the system.
+# Needed for server launch
+bind
+listen
+# Needed by server upon client connect
+accept4
```
Comment (jdstrand) [inline interfaces/builtin/wayland.go]: Pease add `accept` back for the architectures that use it instead of `accept4`.

File: interfaces/builtin/wayland.go [hunk 13]
```diff
@@ -19,32 +19,127 @@
...
+# for udev
+socket AF_NETLINK - NETLINK_KOBJECT_UEVENT
+`
+
+const waylandConnectedPlugAppArmor = `
+# Allow access to the Wayland compositor server socket
 owner /run/user/*/wayland-[0-9]* rw,
 
+# XWayland needs access to this
+/run/xwayland-shared-* rw,
```
Comment (jdstrand) [inline interfaces/builtin/wayland.go]: You removed this from waylandPermanentSlotAppArmor, but left it here. This should be removed too, no?

File: interfaces/builtin/wayland.go [hunk 14]
```diff
@@ -19,32 +19,127 @@
...
+	}
+	return nil
+}
+
+func (iface *waylandInterface) UDevPermanentSlot(spec *udev.Specification, slot *snap.SlotInfo) error {
+	spec.TagDevice(`KERNEL=="tty[0-9]*"`)
+	spec.TagDevice(`KERNEL=="mice"`)
+	spec.TagDevice(`KERNEL=="mouse[0-9]*"`)
+	spec.TagDevice(`KERNEL=="event[0-9]*"`)
+	spec.TagDevice(`KERNEL=="ts[0-9]*"`)
```
Comment (jdstrand) [inline interfaces/builtin/wayland.go]: These should be `if !release.OnClassic {` too I think.

File: interfaces/builtin/wayland.go [hunk 15]
```diff
@@ -19,32 +19,127 @@
...
+
+# needed since Wayland is a display server and needs to configure tty devices
+capability sys_tty_config,
+/dev/tty[0-9]* rw,
+
+# Create the Wayland socket and lock file
+/run/user/[0-9]*/wayland-[0-9]* rw,
+
+# Allow write access to create XDG_RUNTIME_DIR (until lp:1656340 is fixed)
+/run/user/[0-9]*/ w,
```
Comment (jdstrand) [inline interfaces/builtin/wayland.go]: LP: #1656340 was about not creating `/run/user/<uid>/snap.$SNAP`, not `/run/user/<uid>`. The session manager is supposed to create this directory, not the snap. Why did you add this rule?

File: interfaces/builtin/wayland.go [hunk 16]
```diff
@@ -19,32 +19,127 @@
...
+listen
+# Needed by server upon client connect
+accept4
+# for udev
+socket AF_NETLINK - NETLINK_KOBJECT_UEVENT
+`
+
+const waylandConnectedPlugAppArmor = `
+# Allow access to the Wayland compositor server socket
 owner /run/user/*/wayland-[0-9]* rw,
```
Comment (jdstrand) [inline interfaces/builtin/wayland.go]: This should be:
```
owner /run/user/[0-9]*/wayland-[0-9]* rw,
```

I think you will also need:
```
owner /run/user/[0-9]*/wayland-shared-* rw,
owner /run/user/[0-9]*/wayland-cursor-shared-* rw,
```

File: interfaces/builtin/wayland.go [hunk 17]
```diff
@@ -19,32 +19,127 @@
...
+const waylandPermanentSlotAppArmor = `
+# Description: Allow operating as a Wayland display server. This gives privileged access
+# to the system.
+
+# needed since Wayland is a display server and needs to configure tty devices
+capability sys_tty_config,
+/dev/tty[0-9]* rw,
+
+# Create the Wayland socket and lock file
+/run/user/[0-9]*/wayland-[0-9]* rw,
```
Comment (jdstrand) [inline interfaces/builtin/wayland.go]: I think you will need:
```
/run/user/[0-9]*/wayland-shared-* rw,
/run/user/[0-9]*/wayland-cursor-shared-* rw,
```

File: interfaces/builtin/wayland.go (hunk 15 above)
Comment (gerboland) [reply interfaces/builtin/wayland.go]: Mir is running as root, and I'm seeing /run/user/0 does not exist.  Well with
        [ -n "$XDG_RUNTIME_DIR" ] && mkdir -p $XDG_RUNTIME_DIR -m 700
I get 
audit[6690]: AVC apparmor="DENIED" operation="mkdir" profile="snap.mir-kiosk.mir-kiosk" name="/run/user/0/" pid=6690 comm="mkdir" requested_mask="c" denied_mask="c" fsuid=0 ouid=0
mir-kiosk.mir-kiosk[6286]: mkdir: cannot create directory ‘/run/user/0’: Permission denied

Session Manager bug?

File: interfaces/builtin/wayland.go (hunk 13 above)
Comment (gerboland) [reply interfaces/builtin/wayland.go]: I'm building my Chromium snap to incorporate XWayland, which will create that socket on startup. I was thinking it handiest to add support for any XWayland client to the Wayland interface. But in retrospect that's not very secure.

I'll iterate this a bit more.

File: interfaces/builtin/wayland.go (hunk 14 above)
Comment (gerboland) [reply interfaces/builtin/wayland.go]: Including tty too?

File: interfaces/builtin/wayland.go (hunk 15 above)
Comment (jdstrand) [reply interfaces/builtin/wayland.go]: In this case I assume you mean that you are starting mir as a daemon. In this case, there is no user session and IMO it would be snappy's responsibility to create this directory. We had a previous PR for this in snap-confine, but it was never committed for various reasons. @zyga - can you comment? IIRC, the issue was the parent directories of `/run/user/<uid>` but I think if we just create `/run/user/<uid>` as 0700 as per the specification and log (perhaps not fail) if `/run/user` does not exist, we should be fine.

File: interfaces/builtin/wayland.go (hunk 14 above)
Comment (jdstrand) [reply interfaces/builtin/wayland.go]: Yes. Note that this is the UDevPermanentSlot policy-- ie, what to tag for the slot provider so that the slot provider has these devices added to its device cgroup so it may function. When OnClassic, weston, mutter, etc are not running under snappy and are implicitly provided by the classic distro, and therefore there is no 'slot provider snap' where we need to tag these devices.

File: interfaces/builtin/wayland.go (hunk 15 above)
Comment (gerboland) [reply interfaces/builtin/wayland.go]: Correct assumption. Daemonized Mir just needs an accessible XDG_RUNTIME_DIR somehow.

File: interfaces/builtin/wayland.go (hunk 14 above)
Comment (gerboland) [reply interfaces/builtin/wayland.go]: Thanks for the explanation. Learning lots..

File: interfaces/builtin/wayland.go (hunk 13 above)
Comment (gerboland) [reply interfaces/builtin/wayland.go]: Yes you right, this is pointless, removing.

File: interfaces/builtin/wayland.go (hunk 17 above)
Comment (gerboland) [reply interfaces/builtin/wayland.go]: I've been tracking these down, these are files for shared memory created by wayland  (another one being xwayland-shared), to pass data between client and compositor, for example pass pixel buffers, cursor images.

The tricky thing is that clients will create them in their XDG_RUNTIME_DIR. For a snapped app, this directory is per-snap. Non-snapped apps use the normal /run/user/`id -u`

It is correct that the Wayland socket is in a shared place, so multiple clients can connect to it. But these shared memory files I think can be confined inside the snap's XDG_RUNTIME_DIR, as I don't believe they are ever shared between clients, they are purely for client-server communication.

For snapped apps, I think in addition to your proposal we should add

      /run/user/[0-9]*/snap.*/wayland-shared-* rw,
      /run/user/[0-9]*/snap.*/wayland-cursor-shared-* rw,
      /run/user/[0-9]*/snap.*/xwayland-shared-* rw,

(XWayland clients create the later). WDYT?

Comment (jdstrand) [reply interfaces/builtin/wayland.go]: It would be fine for the server to do this, but note that @jhenstridge is working on user mounts. This will allow the snap's XDG_RUNTIME_DIR to be /run/user/`id -u` on the inside, but bind mounted on top of a snap-specific directory, then snapd/snap-confine/snap-update-ns would bind mount the wayland socket into the snap-specific dir. This would mean we could get rid of the desktop helper bits for the compat symlink (this also more closely follows the upstream designs). 

How these intersect with the above accesses... I'm not sure. With blackbox testing, it seems to be initiated on the client side (they do an open() then unlink()) in the client's XDG_RUNTIME_DIR and I think the client must pass the fd over to the server. Ie, if I add an appropriate 'audit deny' rule, I can see denials like this:

```
apparmor="DENIED" operation="mknod" profile="snap.gedit.gedit" name="/run/user/1000/snap.gedit/wayland-cursor-shared-Hcwp8Q" pid=4896 comm="gedit" requested_mask="c" denied_mask="c" fsuid=1000 ouid=1000
```

as such, what you should do is add waylandConnectedSlotAppArmor policy, like so:
```
const waylandConnectedSlotAppArmor = `
# Allow access to client Wayland sockets for connected snaps
owner /run/user/[0-9]*/###PLUG_SECURITY_TAGS###/wayland-shared-* rw,
owner /run/user/[0-9]*/###PLUG_SECURITY_TAGS###/wayland-cursor-shared-* rw,
owner /run/user/[0-9]*/###PLUG_SECURITY_TAGS###/xwayland-shared-* rw,
```

This will allow the server to access these sockets for connected snaps. If my blackbox testing analysis is correct, this should continue to work after user mounts are in effect, because the slot side will continue to see these in /run/user/`id -u`/snap.*.

Comment (jdstrand) [reply interfaces/builtin/wayland.go]: In addition to the above, the waylandPermanentSlotAppArmor policy would need this to handle unconfined applications:
```
owner /run/user/[0-9]*/wayland-shared-* rw,
owner /run/user/[0-9]*/wayland-cursor-shared-* rw,
owner /run/user/[0-9]*/xwayland-shared-* rw,
```

**Comment by gerboland** [comment]
Assuming tests pass on CI, this should be good for another review pass. Thanks

File: interfaces/builtin/wayland_test.go [hunk 18]
```diff
@@ -64,33 +80,138 @@ func (s *WaylandInterfaceSuite) TestName(c *C) {
...
 	// connected plug to core slot
 	spec := &apparmor.Specification{}
 	c.Assert(spec.AddConnectedPlug(s.iface, s.plug, s.coreSlot), IsNil)
 	c.Assert(spec.SecurityTags(), DeepEquals, []string{"snap.consumer.app"})
-	c.Assert(spec.SnippetForTag("snap.consumer.app"), testutil.Contains, "owner /run/user/*/wayland-[0-9]* rw,")
+	c.Assert(spec.SnippetForTag("snap.consumer.app"), testutil.Contains, "/run/user/[0-9]*/wayland-[0-9]* rw,")
+	c.Assert(spec.SnippetForTag("snap.consumer.app"), testutil.Contains, "owner /run/user/[0-9]*/wayland-shared-* rw,")
+	c.Assert(spec.SnippetForTag("snap.consumer.app"), testutil.Contains, "owner /run/user/[0-9]*/wayland-cursor-shared-* rw,")
+	c.Assert(spec.SnippetForTag("snap.consumer.app"), testutil.Contains, "owner /run/user/[0-9]*/xwayland-shared-* rw,")
+	c.Assert(spec.SnippetForTag("snap.consumer.app"), testutil.Contains, "/etc/drirc r,")
```
Comment (jdstrand) [inline interfaces/builtin/wayland_test.go]: Typically you only need to test for one rule that is not shared with the other snippets, eg, here the drirc rule would work. Sometimes it is nice to also test other types of rules (eg, the security label on unix sockets, snap-specific file paths, etc. This applies to the ConnectedSlot and PermanentSlot policy below too. Not a blocker.

File: interfaces/builtin/wayland.go [hunk 19]
```diff
@@ -19,32 +19,157 @@
...
+# Allow access to common client Wayland sockets for connected snaps
+owner /run/user/[0-9]*/###PLUG_SECURITY_TAGS###/wayland-shared-* rw,
+owner /run/user/[0-9]*/###PLUG_SECURITY_TAGS###/wayland-cursor-shared-* rw,
+owner /run/user/[0-9]*/###PLUG_SECURITY_TAGS###/xwayland-shared-* rw,
 `
 
 const waylandConnectedPlugAppArmor = `
-# Description: Can access compositors supporting the wayland protocol
+# Allow access to the Wayland compositor server socket
+/run/user/[0-9]*/wayland-[0-9]* rw,
```
[the remark this answers is not in the document]
Comment (gerboland) [reply interfaces/builtin/wayland.go]: This ends up in the client's AppArmor policy, yeah? I didn't add owner as I was leaving open the possibility that the wayland socket may have different uid/gid than the client. Being too liberal?

File: interfaces/builtin/wayland.go [hunk 20]
```diff
@@ -19,32 +19,157 @@
...
 
-# Allow access to the wayland compsitor server socket
-owner /run/user/*/wayland-[0-9]* rw,
+# Allow client create common Wayland sockets to share with compositor
+owner /run/user/[0-9]*/wayland-shared-* rw,
+owner /run/user/[0-9]*/wayland-cursor-shared-* rw,
+owner /run/user/[0-9]*/xwayland-shared-* rw,
```
[the remark this answers is not in the document]
Comment (gerboland) [reply interfaces/builtin/wayland.go]: Yes you're totally right.

File: interfaces/builtin/wayland.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (jdstrand) [reply interfaces/builtin/wayland.go]: It is in the client's apparmor policy, yes. AIUI, part of the design of wayland was to not need for it to run as root (like X required for a long time) and indeed, mutter, plasma/wayland and sway all run their compositors under the uid of the user, which is why I thought this could use 'owner' match. What is the use case for a client running under one uid to use a wayland socket created/used by a compositor running as a different user?

File: interfaces/builtin/wayland_test.go [hunk 21]
```diff
@@ -139,10 +120,6 @@ func (s *WaylandInterfaceSuite) TestAppArmorSpecOnClassic(c *C) {
 	c.Assert(spec.AddConnectedPlug(s.iface, s.plug, s.classicSlot), IsNil)
 	c.Assert(spec.SecurityTags(), DeepEquals, []string{"snap.consumer.app"})
 	c.Assert(spec.SnippetForTag("snap.consumer.app"), testutil.Contains, "/run/user/[0-9]*/wayland-[0-9]* rw,")
```
Comment (jdstrand) [inline interfaces/builtin/wayland_test.go]: Since you used 'owner' above now, it would be nice to adjust this for owner so that if we change the above rule to remove it, this would fail.

Comment (gerboland) [reply interfaces/builtin/wayland_test.go]: Done. Sorry I missed that.
Thanks for the patient review!

**Comment by gerboland** [comment]
CI failing on fedora 26 setup...
